### PR TITLE
52 impersonate other users

### LINKED ISSUE #5: impersonate other users (of other types)

### PR DESCRIPTION
**Pull request by bjorvack**
closes #5

### CODE REVIEW

File: Security/ObjectUserProvider.php [hunk 1]
```diff
@@ -2,31 +2,41 @@
...
-        $user = $this->userRepository->findByUsername($username);
+        $user = null;
+
+        foreach ($this->userRepositoryCollection->all() as $repository) {
+            $user = $repository->findByUsername($username);
+
+            if ($user instanceof User) {
+                break;
+            }
+        }
```
Comment (WouterSioen) [inline Security/ObjectUserProvider.php]: why not return instead of break?

Comment (bjorvack) [inline Security/ObjectUserProvider.php]: fixed 38e080e

File: Resources/doc/index.md [hunk 2]
```diff
@@ -46,19 +46,21 @@ security:
 
 To create a user provider, you need two services:
 
-* A UserRepository implementing our
+* A UserRepositoryCollection with UserRepositories implementing our
 SumoCoders\FrameworkMultiUserBundle\User\UserRepository interface.
 * An instance of the ObjectProvider getting the repository as argument
 
 ```yaml
 # app/config/confing.yml
```
Comment (StijnVrolijk) [inline Resources/doc/index.md]: tis niet uw code maar er staat een typo in `config.yml`

Comment (bjorvack) [inline Resources/doc/index.md]: fixed b4439c7

**Comment by StijnVrolijk** [comment]
:shipit: @sumocoders/php-team kijken jullie nog eens of er niets over het hoofd gezien is?

**Comment by carakas** [comment]
er zijn blijkbaar wel nog conflicts

File: Resources/doc/index.md [hunk 3]
```diff
@@ -136,3 +136,14 @@ services:
...
+
+## User impersonation
+
+Add a `switch_user` to the firewall to enable user impersonation
+
+```yaml
+security:
+    firewalls:
+        main:
+          switch_user: { role: ROLE_ALLOWED_TO_SWITCH, parameter: _switch_user }
```
Comment (carakas) [inline Resources/doc/index.md]: why the _ before switch_user

File: Security/ObjectUserProvider.php [hunk 4]
```diff
@@ -2,31 +2,41 @@
...
+        $this->userRepositoryCollection = $userRepositoryCollection;
     }
 
     /**
      * {@inheritDoc}
      */
     public function loadUserByUsername($username)
     {
-        $user = $this->userRepository->findByUsername($username);
+        $user = null;
```
Comment (carakas) [inline Security/ObjectUserProvider.php]: because you return the user from within the foreach you don't need this anymore

Comment (carakas) [inline Security/ObjectUserProvider.php]: you can just return null if you come out of the foreach

File: Security/ObjectUserProvider.php [hunk 5]
```diff
@@ -56,6 +66,12 @@ public function refreshUser(UserInterface $user)
      */
     public function supportsClass($class)
     {
-        return $this->userRepository->supportsClass($class);
+        try {
```
Comment (carakas) [inline Security/ObjectUserProvider.php]: isn't there an other way than exception driven development to check this ?

File: Resources/doc/index.md (hunk 3 above)
Comment (bjorvack) [inline Resources/doc/index.md]: is the default in symfony, could be changed to switch_user
http://symfony.com/doc/current/cookbook/security/impersonating_user.html

Comment (carakas) [inline Resources/doc/index.md]: if it is the default I would leave it like that

File: Security/ObjectUserProvider.php (hunk 4 above)
Comment (bjorvack) [inline Security/ObjectUserProvider.php]: isn't it better to throw an exception when no user is found

File: Security/ObjectUserProvider.php (hunk 5 above)
Comment (bjorvack) [inline Security/ObjectUserProvider.php]: $this->userRepositoryCollection should always have a repository for the correct user type,
if it doesn't exist it throws an RepositoryNotRegisteredException

File: Security/ObjectUserProvider.php (hunk 4 above)
Comment (carakas) [inline Security/ObjectUserProvider.php]: yes it is, but you still can remove $user = null;

Comment (carakas) [inline Security/ObjectUserProvider.php]: the diff didn't show the exception after this

Comment (bjorvack) [inline Security/ObjectUserProvider.php]: fixed b73409a

File: Security/ObjectUserProvider.php (hunk 5 above)
Comment (bjorvack) [inline Security/ObjectUserProvider.php]: fixed dacf71d

File: User/UserRepositoryCollection.php [hunk 6]
```diff
@@ -71,6 +71,22 @@ public function findRepositoryByClassName($className)
...
+     * @return bool
+     */
+    public function supportsClass($className)
+    {
+        foreach ($this->userRepositories as $repository) {
+            if ($repository->supportsClass($className)) {
+                return true;
+            }
+        }
+    }
```
Comment (StijnVrolijk) [inline User/UserRepositoryCollection.php]: Er ontbreekt nog een `return false`

Comment (bjorvack) [inline User/UserRepositoryCollection.php]: fixed b4a6e47

**Pull request merged**